### PR TITLE
Expand parallel axis race condition checks in block merging algorithm

### PR DESCRIPTION
**Pull request by jdahm**
## Description

Replaces `has_disallowed_write_after_read_in` and `has_disallowed_read_after_write_in` with a specific check `has_disallowed_read_with_offset_and_write` that:
1. Splits if an API field would both be read with an offset and written to in the same computation block.
2. Splits if parallel computation blocks and any field is both read with an extent in the sequential axis and written to.

Also adds the same type of check to the StageMergingWrapper for all fields (not just API).

The specific changes above with respect to what existed before:
- `MultistageMergingWrapper`: Read with offset after write is checked for parallel axes on API fields (may cause more splitting)
- `MultistageMergingWrapper`: Temporary fields are ignored in the parallel axes check (may cause more merging)
- `MultistageMergingWrapper`: Checks offsets more accurately by not accumulating with min/max before adding compute extent shift
- `StageMergingWrapper`: Checks read with offset before/after write on parallel axes for all fields (may cause more splitting)

This implements the write after read condition described in [GridTools/concepts#39](https://github.com/GridTools/concepts/pull/39).

### CODE REVIEW

File: src/gt4py/analysis/passes.py
Comment (jdahm) [inline src/gt4py/analysis/passes.py:663]: Unfortunately the `FieldDependencyGraphCreator` IR visitor does not work with the infos, so this "hack" (or something like it) is necessary.

**Comment by gronerl** [comment]
Tox tests should pass after merging #369

File: src/gt4py/analysis/passes.py
Comment (jdahm) [inline src/gt4py/analysis/passes.py:544]: This logic depends on the visitor visiting the assignments in the order they appear. I'm not sure that's a requirement of the IRVisitor, but it is how it has been written.

Comment (gronerl) [reply src/gt4py/analysis/passes.py:544]: While it's nowhere explicitly documented, I think this is a crucial assumption for many of the visitors and passes. I wondered that too at some point, so far it's never not worked :sweat_smile:

File: tests/test_unittest/test_merge_blocks_pass.py
Comment (gronerl) [inline tests/test_unittest/test_merge_blocks_pass.py:54]: Whoever wrote this test seems to have the opinion that this would be wrong. I don't see why since there are no offsets here. Maybe it was @DropD ?

Comment (DropD) [reply tests/test_unittest/test_merge_blocks_pass.py:54]: As the comment says: the read on `inout` happens with extended compute domain, the write does not. According to my understanding of the parallel model at the time of writing this, merging them is not allowed.

Comment (DropD) [reply tests/test_unittest/test_merge_blocks_pass.py:54]: The test came from here: https://github.com/GridTools/gt4py/issues/143#issuecomment-686455071.
Since it was written by Hannes after looking at the parallel model in-depth and confirmed by multiple people I think it requires an explanation why the stages should be allowed to merge.

Comment (fthaler) [reply tests/test_unittest/test_merge_blocks_pass.py:54]: I can confirm that merging this indeed introduces a race condition in GridTools (at least if I interpret the IR here correctly, at which I have never looked so far). The second statement (`out = tmp[some ij offset]`) increases the compute extent. When fusing the last statement (`inout = in`) we get race conditions on inout due to the extended compute domain around each computational block.

Comment (fthaler) [reply tests/test_unittest/test_merge_blocks_pass.py:54]: Maybe worth to mention: this race condition thus only appears if the domain is large enough, that is, larger than a single block (the default block size depends on the GT backend).

Comment (jdahm) [reply tests/test_unittest/test_merge_blocks_pass.py:54]: ah, is it because these end up a single multistage with different stages, and there is no ordering guarantee on them?

Comment (fthaler) [reply tests/test_unittest/test_merge_blocks_pass.py:54]: Yes, exactly. Each multistage means (approximately) one kernel call (kernel == global loop, on the GPU equivalent to a GPU kernel). The stages within the multistage are executed block-wise in this kernel. So between the stages, only block-local synchronization happens. The extreme case is a single-threaded CPU execution, where actually one block is processed after the other. Global synchronization as required in this example (because you read remote-block data from inout which is also the output field) is obviously not possible there.

File: src/gt4py/analysis/passes.py
Comment (gronerl) [inline src/gt4py/analysis/passes.py:605]: You posted on slack that in FV3 you have a performance problem with this approach. The number of cycles can scale combinatorially. I have the following observation:

Effectively what you're doing here is just checking  if any edge that is in a cycle reads at an offset. there is absolutely no need to iterate over all cycles here, you're checking the same edges over and over again (they can be in multiple cycles).

I think there should be an at most O(V*E) algorithm to find the edges you want (e.g. starting a DFS at every node and see which paths lead back to the original node), maybe there is even an O(V+E) algorithm (at least that's the complexity of `networkx.algorithms.components.strongly_connected_components` i'm haven't convinced myself yet but that could be just what you need.)

Comment (gronerl) [reply src/gt4py/analysis/passes.py:605]: Also, this is run every time you try to merge, so you could improve on the runtime by instead of `race_fields.append` directly `return False`. Further, there might be a way to speed things up by using the knowledge that the edges within an existing multistage are fine

Comment (gronerl) [reply src/gt4py/analysis/passes.py]: So I have convinced myself now  and pushed all suggestions except the last point. Let me know if this solves the issue.

Comment (jdahm) [reply src/gt4py/analysis/passes.py]: @gronerl All great points: I should have immediately escaped instead of adding them to the list. Your changes look really good, and that unblocked the FV3 stencil compilation. Thanks again!

File: tests/test_unittest/test_merge_blocks_pass.py
Comment (jdahm) [reply tests/test_unittest/test_merge_blocks_pass.py]: Thanks for the explanation! Off topic: I've always wondered how extended stages are implemented in GridTools. Is the CUDA grid block size covering the largest stage extent and threads are masked out (if condition) every time there's a smaller stage?

**Comment by jdahm** [comment]
After thinking about this today for a while, I realized that due to a few misunderstandings about how the merging works, I was actually able to re-use the existing infrastructure very well. In fact, there is only one more check necessary, and this PR adds it.

This adds a check there is no read with offset on a parallel axis after writing to an API field (note that for temporaries this check does not apply). Before this, there was only a check on all fields that there was no read with sequential axis offset after write.

Does this sound correct?

**Comment by fthaler** [comment]
> After thinking about this today for a while, I realized that due to a few misunderstandings about how the merging works, I was actually able to re-use the existing infrastructure very well. In fact, there is only one more check necessary, and this PR adds it.
> 
> This adds a check there is no read with offset on a parallel axis after writing to an API field (note that for temporaries this check does not apply). Before this, there was only a check on all fields that there was no read with sequential axis offset after write.
> 
> Does this sound correct?

Sounds good. One open PR in the parallel model description actually addresses exactly this issue: https://github.com/GridTools/concepts/pull/39

File: tests/test_unittest/test_merge_blocks_pass.py
Comment (fthaler) [reply tests/test_unittest/test_merge_blocks_pass.py]: That’s basically how it’s working yes. But there’s might be one surprise: the threads within a single block are ordered in a somewhat special way to get better alignment on memory accesses and also slightly less warp divergence. Though, not sure how much this matters on latest GPU architectures…

**Comment by jdahm** [comment]
I just pushed one more change that expands things and cleans it up, but the commit is failing on some tests, so it still needs work. Perhaps if one of you has a chance to look today you can catch my mistake.

File: tests/test_unittest/test_merge_blocks_pass.py
Comment (jdahm) [reply tests/test_unittest/test_merge_blocks_pass.py]: @fthaler - Thanks! I came to the same conclusion as you that this pattern creates incorrect results when `tmp = inout` is in the same multistage as
```
out = tmp[1, 0, 0]
inout = in
```

When you have a moment, I have a few follow-up questions relating to this test:

```
def example(inp: Field[float], out: Field[float], inout: Field[float]):
    with computation(PARALLEL), interval(...):
        tmp = inout
        out = tmp[1, 0, 0]
        inout = inp
```

**Non-scalar temporary fields**

Naively, I would conclude that since `tmp` is not an API field (rather, it is a temporary field) that putting these into the same multistage should not incur a data race since each CUDA block could be using its own shared memory for the temporary field. At least, this is how I would try to code it in a GPU implementation.

However, when I look at the GridTools C++ generated by GT4Py, it seems like `tmp` is being defined globally instead of at the multi-stage level. For example GT v1:
```
using p_tmp = gt::tmp_arg<3, typename storage_traits<float64_t, 0, 1, 1, 1>::store_t>;
...
    computation_t gt_computation = gt::make_computation<backend_t>(
        make_grid(domain),
        gt::make_multistage(gt::execute::parallel(),
            gt::make_stage<stage__10_func>(
                p_inout(), p_tmp()
            )
        ),
        gt::make_multistage(gt::execute::parallel(),
            gt::make_stage<stage__13_func>(
                p_tmp(), p_inp(), p_inout(), p_out()
            )
        )
    );
```
and GT v2:
```
      auto GTComputationCall_111 = [](auto inout, auto inp, auto out) {
        GT_DECLARE_TMP(double, tmp);
        return multi_pass(
            execute_parallel()
                .stage(HorizontalExecution_62(), tmp, inout)
                .stage(HorizontalExecution_63(), out, tmp, inout, inp));
      };
```

Since this is at the top-level and could be used across multi-stages I assume it needs to be allocated in GPU global memory and there's no room for optimization? Or, is it still using shared memory in certain instances?

I'm wondering if there is benefit to distinguishing between temporaries that are used within a multistage and those that could be shared across them. Is GridTools may be doing some of this analysis behind the scenes?

**Stage splitting**

The `gtx86` backend executes the example correctly, but the gtc:gt:cpu_*` backends fail (off topic, but that needs to be fixed). The reason for this is that the last two statements are put in the same stage:
```
struct HorizontalExecution_63 {
  using out = inout_accessor<0, extent<0, 0, 0, 0, 0, 0>>;
  using tmp = in_accessor<1, extent<0, 1, 0, 0, 0, 0>>;
  using inout = inout_accessor<2, extent<0, 0, 0, 0, 0, 0>>;
  using inp = in_accessor<3, extent<0, 0, 0, 0, 0, 0>>;

  using param_list = make_param_list<out, tmp, inout, inp>;

  template <typename Evaluation>
  GT_FUNCTION static void
  apply(Evaluation eval, gridtools::stencil::core::interval<
                             gridtools::stencil::core::level<0, 1, 1>,
                             gridtools::stencil::core::level<1, -1, 1>>) {

    eval(out(0, 0, 0)) = eval(tmp(1, 0, 0));
    eval(inout(0, 0, 0)) = eval(inp(0, 0, 0));
  }
};
```

I'm wondering what is going on under the hood here. To me it looks like an iteration over the un-extended stencil compute domain, with an extent on tmp. Does it have to be in two stages because `tmp` was written to in an extended stage (inside a different multistage)?

**Multistage/Stage Merging Algorithm**

Evidence like this suggests that the block merging algorithm currently implemented in `master` is correctly splitting to multistages and stages to prevent race conditions. Have you identified any areas where there are unnecessary splits? Said differently, are the requirements in the code exactly what GridTools has specified, or are they overly conservative while yielding correct results?

**Outro**

If you made it this far, thanks for parsing through it :) Perhaps it would be easier to setup a meeting early next week to discuss this. Larger stencils in our dycore seem to have incorrect results with the GT backends (mostly or always combined with horizontal regions), and I suspect it's related to not fully understanding this model. We will need to fully understand to get the semantics of block merging correct with the introduction of horizontal ifs.

cc: @mbianco @anstaf

**Comment by jdahm** [comment]
After being thoroughly confused about this on Friday, after some digging and finding a bug in this implementation that had caused confusion, this is ready for review!

File: src/gt4py/analysis/passes.py
Comment (jdahm) [inline src/gt4py/analysis/passes.py:574]: This can and should be refactored to use a generator function, but the core problem is that the infos are computing the input and output extents in not quite the right way. They are `OR`-ing each usage, which is I think more careful but less accurate than what I propose doing, which is shifting it by the compute extents then `OR`-ing only within the IJ block. Could you verify that this is indeed what we want?

If it is what we should be doing, then I propose we move this logic into the correct places in `BlockMaker.visit_ComputationBlock` and `MultiStageMergingWrapper.merge_with`.

**Comment by jdahm** [comment]
I'm happy with how this looks and it seems to be passing all tests (here and in our dynamical core) so it's ready for another review when you have time.

**Comment by gronerl** [comment]
bors try

**Comment by gronerl** [comment]
bors try

**Comment by gronerl** [comment]
bors try-
bors try

**Review by gronerl — APPROVED** [review]
Looks good to me.

**Pull request merged**
